File: 98. Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode pre = null;

    public boolean isValidBST(TreeNode root) {
        if(root == null) {
            return true;
        }
        boolean left = isValidBST(root.left);
        if(pre != null && root.val <= pre.val) {
            return false;
        }
        pre = root;
        boolean right = isValidBST(root.right);
        return left && right;
    }
}

File: 98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // Version 1
    // vector<int> arr;

    // void traversal(TreeNode* root) {
    //     if(root == nullptr) {
    //         return;
    //     }
    //     traversal(root->left);
    //     arr.push_back(root->val);
    //     traversal(root->right);
    // }

    // bool isValidBST(TreeNode* root) {
    //     traversal(root);
    //     for(int i = 1; i < arr.size(); i++) {
    //         if(arr[i] <= arr[i - 1]) {
    //             return false;
    //         }
    //     }
    //     return true;
    // }

    // Version 2
    TreeNode* pre = nullptr;
    bool isValidBST(TreeNode* root) {
        if(root == nullptr) {
            return true;
        }
        bool left = isValidBST(root->left);
        if(pre != nullptr && root->val <= pre->val) {
            return false;
        }
        pre = root;
        bool right = isValidBST(root->right);
        return left && right;
    }
};